### ISSUE TITLE
Bug: [@typescript-eslint/no-this-alias] false positive with a generator

### ISSUE THREAD

**Issue by vadzim** [typescript-eslint/typescript-eslint#6964]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=MYGwhgzhAECyCeBhcVoG8BQ1oAcBOAlgG5gAuAptBAPYC25AamCAK7kBc0Adi7QEbk8AbgxZowalwik8LYKWp4AFAEp0Y7KQAWBCADoa9Jq0oBeaABYATCOwBfUdhws+IAsGi14AMRZd5BJKq6tjYElKkVOQgAGbQ5tq6QtAA9CnQACo6MG5clDIEAOaFgjDalADkXNQAtIkQNcwEkBXQsiDkeo6hMX4BkgBU0ARceXgA4uRjZIrBmKGh8ATRACZRsQZ0jMxstqEOGm3kpCx4XMOjgpPTCsoqew4OQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1vwAtLlaAQ0aDUGSImjQO0SODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
class MyClass {
  private someValue: number;

  constructor() {
    this.someValue = 42;
  }

  public myFunction() {
    const self = this; // This line triggers the 'no-this-alias' rule.

    function* innerGenerator() {
      yield self.someValue;
    }

    return innerGenerator();
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-this-alias": "error"
  }
}
```

### Expected Result

I expected the 'no-this-alias' rule not to be triggered in this scenario, as using the "this" alias (`const self = this`) is necessary to access the `someValue` property from within the inner generator function. The code should be considered valid without any warnings or errors related to the 'no-this-alias' rule.

### Actual Result

The 'no-this-alias' rule is triggered on the line `const self = this;`, producing the following warning/error message:
```
error  @typescript-eslint/no-this-alias: Unexpected aliasing of 'this'.  no-this-alias
```
This rule violation is considered a false positive, as the use of the "this" alias is necessary for the correct functioning of the generator function.

**Comment by Josh-Cena** [comment]
The rule errors on creating aliases for `this`; it cannot check where that alias is used because doing so is very complicated (it has to go through every child scope; and what if this value is returned from the function?). I suggest using an ignore comment here instead.

**Comment by vadzim** [comment]
Thank you for the suggestion. I understand that using an ignore comment is a temporary workaround. I'm not familiar enough with the codebase to determine how much detecting generators might affect performance. However, I believe it would be worth investigating if a fix can be implemented without significantly impacting performance.

**Comment by bradzacher** [comment]
This is working as expected - as Josh said the rule simply reports on any aliasing of `this`, regardless of how its used.

This isn't *strictly* necessary, as I believe you can `.bind` your function's `this` context.

If you choose to stylistically use aliased `this` instead, then you should use a disable comment.
